### PR TITLE
Created Searchplugin.java with default implementation 

### LINKED ISSUE #331: [FEATURE] Create SearchExtension interface
**Is your feature request related to a problem?**

Extension points (see #315) will be (optionally) implemented on Extensions. Eventually these points may need to be communicated to OpenSearch to be registered in the appropriate service or module.

Allowing extensions to inherit from appropriate interfaces depending on their type will match the current Plugin implementation.

**What solution would you like?**

Add a new interface SearchExtension with default implementations for extension points matching SearchPlugin.java.

### PR DESCRIPTION
**Pull request by varuntumbe**
Signed-off-by: varuntumbe <varuntumbe1@gmail.com>

### Description
Add a new interface SearchExtension with default implementations for extension points matching SearchPlugin.java.

Copy pasted the default implementation of SearchPlugin.java to SearchExtension.java 

### Issues Resolved
Fixes #331

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by varuntumbe** [comment]
Tell me If I have to add anything

Thanks and Regards
Varun

**Review by dbwiddis — COMMENTED** [review]
Thanks for this PR, @varuntumbe!

The code looks good.  There are some formatting tweaks (extra spaces, need a license header, need a newline at the end of file) that you should be able to fix by running `./gradlew spotlessApply` and committing the results.

**Comment by varuntumbe** [comment]
Yes I applied the spotlessApply and RequireJavaDoc and Documentation linkage was giving build errors. I fixed it as of now temporarily. But  still it throws some warnings as no @return for methods.

Tell If anything else has to be done to close this, I will do one more of this #323 ( I already copied and just there to commit ) and next I will proceed to #319 If its okay.


Thanks and Regards
Varun

**Comment by dbwiddis** [comment]
> Yes I applied the spotlessApply and RequireJavaDoc and Documentation linkage was giving build errors. I fixed it as of now temporarily. But still it throws some warnings as no @return for methods.

Yeah, you need `@param` and `@return` for constructors with parameters.   I think it's not required on the getters with same name as the field they are getting.

The one `@link` you removed goes to a not-yet-written `Extension` interface extension point, so for now to preserve the text could you just change what was `@link` to `@code` and we can put it back to link later?

File: src/main/java/org/opensearch/sdk/SearchExtension.java [hunk 1]
```diff
@@ -0,0 +1,860 @@
...
+         * TODO: Doc has to be written for the below method
+         */
+        public FetchPhaseConstructionContext(Map<String, Highlighter> highlighters) {
+            this.highlighters = highlighters;
+        }
+
+        /**
+         * TODO: Doc has to be written for the below method
+         */
+        public Map<String, Highlighter> getHighlighters() {
```
[the remark this answers is not in the document]
Comment (varuntumbe) [reply src/main/java/org/opensearch/sdk/SearchExtension.java:856]: yes it does not need it for getter, I removed and its building fine

File: src/main/java/org/opensearch/sdk/SearchExtension.java [hunk 2]
```diff
@@ -0,0 +1,860 @@
...
+
+    /**
+     * Context available during fetch phase construction.
+     */
+    class FetchPhaseConstructionContext {
+        private final Map<String, Highlighter> highlighters;
+
+        /**
+         * TODO: Doc has to be written for the below method
+         */
```
[the remark this answers is not in the document]
Comment (varuntumbe) [reply src/main/java/org/opensearch/sdk/SearchExtension.java:848]: Done

File: src/main/java/org/opensearch/sdk/SearchExtension.java [hunk 3]
```diff
@@ -0,0 +1,860 @@
...
+     * The new {@link QueryPhaseSearcher} added by this Extension. At the moment, only one {@link QueryPhaseSearcher} is supported per
+     * instance, the {@link IllegalStateException} is going to be thrown if more then one Extension tries to register
+     * {@link QueryPhaseSearcher} implementation.
+     */
+    default Optional<QueryPhaseSearcher> getQueryPhaseSearcher() {
+        return Optional.empty();
+    }
+
+    /**
+     * The executor service provider to be used at search
```
[the remark this answers is not in the document]
Comment (varuntumbe) [reply src/main/java/org/opensearch/sdk/SearchExtension.java:185]: Yes, I can add that method to Extension Interface If its required.

[the remark this answers is not in the document]
Comment (varuntumbe) [reply src/main/java/org/opensearch/sdk/SearchExtension.java:185]: Sure no problem, I will paste that one

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Thanks for your patience working through this!  A few javadoc nits.

File: src/main/java/org/opensearch/sdk/Extension.java [hunk 4]
```diff
@@ -89,4 +91,15 @@ default List<NamedXContentRegistry.Entry> getNamedXContent() {
     default Map<String, Class<? extends TransportAction<? extends ActionRequest, ? extends ActionResponse>>> getActions() {
         return Collections.emptyMap();
     }
+
+    /**
+     * Provides the list of this plugin's custom thread pools, empty if
```
[the remark this answers is not in the document]
Comment (varuntumbe) [reply src/main/java/org/opensearch/sdk/Extension.java:96]: Oops did not notice this one. Fixed

File: src/main/java/org/opensearch/sdk/SearchExtension.java [hunk 5]
```diff
@@ -844,15 +844,12 @@ class FetchPhaseConstructionContext {
         private final Map<String, Highlighter> highlighters;
 
         /**
-         * TODO: Doc has to be written for the below method
+         * Instantiates this object with the specified @param highlighters
```
[the remark this answers is not in the document]
Comment (varuntumbe) [reply src/main/java/org/opensearch/sdk/SearchExtension.java:847]: Done

**Comment by varuntumbe** [comment]
If there is anything left out, I will do it tomorrow. 

Thanks and Regards
Varun

**Review by dbwiddis — APPROVED** [review]

**Review by joshpalis — APPROVED** [review]
LGTM. Thanks for the effort @varuntumbe.

**Comment by varuntumbe** [comment]
Thanks @joshpalis. I Hope to learn and contribute more in the future

**Pull request merged**